### ISSUE TITLE
[TypeScript]: Comment is not printed ("Please report this error!")

### ISSUE THREAD

**Issue by G-Rath** [prettier/prettier#7574]
**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBDJmAqAPAPQBUmkAtpQjKcQHyYC8mAjANwgA0IEADjACW6ZKGwAnCRADuABUkI0yENgBuEIQBMeIAEYTsYANZwYAZX5GhUAObIYEgK5xeACxiUANgHU3Q+DQrMDhzJQChNQCATxUwNGVeGzQ4CRg5Q1tKbGQAM2wvFN4AKzQADwAhQxMzc2waABkbODyCopBSsvMbWy84AEUnCHhWwtcQKwkUiRU9bD04L11+CRsYH20YN2QADgAGXhWIFJ9DfhUVuGm1Ft4ARyH4DIFlFGw0AFooODgtX90JHAHkJARlsFkckh8mNeClKEIHM5xmgen1BsMWlC2uMYPMNlotsgAEy8RzYIReHoAYQg1EhICg0FuICcKXw81e0PaahcAEkoH9YOYwKtBABBAXmGDRPqjFIAX3lQA)
```sh
--parser babel
```

**Input:**
```jsx
const a: T</* comment*/> = 1;
```

**Output:**
```jsx
Error: Comment "comment" was not printed. Please report this error!
    at https://prettier.io/lib/standalone.js:15543:15
    at Array.forEach (<anonymous>)
    at ensureAllCommentsPrinted (https://prettier.io/lib/standalone.js:15541:17)
    at coreFormat (https://prettier.io/lib/standalone.js:15592:5)
    at format (https://prettier.io/lib/standalone.js:15832:75)
    at formatWithCursor (https://prettier.io/lib/standalone.js:15848:14)
    at https://prettier.io/lib/standalone.js:31794:17
    at Object.format (https://prettier.io/lib/standalone.js:31802:14)
    at formatCode (https://prettier.io/worker.js:234:21)
    at handleMessage (https://prettier.io/worker.js:185:18)
```

**Expected behavior:**

It to not error, and to print the comment.

**Comment by Skalman** [comment]
This is a (syntax?) error in Typescript ([TS playground](https://www.typescriptlang.org/play/#code/MYewdgzgLgBAhgLhgFQDwHoBUNQFtcCmYUm6AfDALwwCMA3AFBA)). But it seems like Babel doesn't report any errors ([Babel playground](https://babeljs.io/repl/#?browsers=&build=&builtIns=false&spec=false&loose=false&code_lz=MYewdgzgLgBAhgLhgFQDwHoBUNQFtcCmYUm6AfDALwwCMA3EA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=es2015%2Creact%2Cstage-2%2Ctypescript&prettier=false&targets=&version=7.8.4&externalPlugins=)). Interestingly the problem happens regardless of whether you use the Typescript or Babel compiler.

The correct solution would be for Prettier to refuse to reformat the code, instead reporting an error.

E.g. the following correctly reformatted.
```ts
const a: T<U/* comment*/> = 1;
```

**Comment by G-Rath** [comment]
@Skalman it's a semantic error, since [TS still parses it](https://astexplorer.net/#/gist/9fb943b2bb9b390743e19213fb09ec00/36eebca7f73d97cecf39b46e551522dad55467d3); as such, generally I'd say that prettier should format it since there's valid AST.

In fact everything still parses it - even `@typescript-eslint/parser` doesn't fail like it typically does.

**Comment by Skalman** [comment]
@G-Rath, so this is a bug in Prettier rather than both Babel and tsc?

TS will parse and deal with [anything](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=1&pc=39#code/JIAgJg9gdg5ALiOALAllA1o1BnELcBuAhgDYpggAqAngA4CmAygMYBOKtcQA), and I don't really understand how errors are categorized.

**Comment by G-Rath** [comment]
@Skalman I would say so - typically two bugs are less likely than one 😉 

> TS will parse and deal with anything,

True, which is because its robust and why prettier hasn't dropped it in favor of babel 🙂 

Generally, it's better for user experience to be as robust and supporting as possible; while it wouldn't be "wrong" to have prettier error, in doing so it means that a file with this code can never be formatted by prettier.

That's actually one of the differences between eslint & typescript: eslint takes the erroring route bailing out at the first non-spec compliant or syntactically invalid character it runs into - the result is that ESLint just won't lint your files if they have such code in them.

Meanwhile TypeScript strives to try and tell you as much about your code as possible, and so presses on when it encouters code that would make eslint error.

Neither is right or wrong - they're serving different purposes, but I think for prettier in cases like these where the price of adding support is relatively cheap, it's worth accepting them as a bug in prettier rather than hard failing 🙂 

------

The above is mostly IMO - I don't speak for the prettier team, nor the eslint or typescript teams, but thats what I feel gives the best UX 🤷‍♂

**Comment by thorn0** [comment]
@Skalman
> The correct solution would be for Prettier to refuse to reformat the code

Not really. Prettier is a code formatter in the first place, not a correction checker. Imagine you're trying to get through some messy code. Help from a formatter would really come in handy, but because of a trivial error Prettier refuses to help. Not good. That's why Prettier uses the Babel parser in the [error recovery mode](https://babeljs.io/blog/2019/11/05/7.7.0#parser-error-recovery-10363-https-githubcom-babel-babel-pull-10363) and benefits from the fact the TS parser is very permissive.

@G-Rath 
> but I think for prettier in cases like these where the price of adding support is relatively cheap, it's worth accepting them as a bug in prettier rather than hard failing 

Exactly.

**Comment by Skalman** [comment]
@thorn0, I had assumed that this was the policy of Prettier, since Prettier does _not_ format code with syntax errors.

Sorry for derailing the discussion. :smile:

**Comment by G-Rath** [comment]
> Sorry for derailing the discussion. 

I would class this as a healthy derailment - it's always good to talk about these things, and we all learnt something from it (hopefully), so no harm no foul 🙂

**Comment by sosukesuzuki** [comment]
Fixed by #7729
